### ISSUE TITLE
Registered users suddenly loses their accounts

### ISSUE THREAD

**Issue by Sopor** [luadch/luadch#189]
It seems that Luadch will suddenly lose registered users.

I have been contacted by several Hub Owners the past month with the same issue and now i can verify that it is happening for me too.

~~It also seems that the `user.tbl.bak` never gets updated (verified on four different hubs).~~ Moved to #191

I'm running 2.24 RC3 on Linux.

**Comment by ryehamstrawberry** [comment]
Posting to confirm this is a fairly major issue that has been plaguing me too and also seemingly affects all account modifications in general.

The following account changes appear to execute successfully but then later (within ~24h or so) the changes revert, from:
- !reg;
- !delreg;
- to !ban.

I have experienced users receiving their account one day then losing them the next to users having their accounts removed then able to join the following day because changes are not saving properly.

Let me know if I can provide any further assistance.

---
Version: Luadch v2.24 [RC2]
OS: Linux

**Comment by Kungen123** [comment]
when will this be fixed ?

**Comment by pulsar-de** [comment]
Unfortunately i don't want to be able to reproduce this. If anyone knows a way to reproduce it please post.

**Comment by Sopor** [comment]
Sometimes it can take week(s) before it happens and sometimes a couple of hours, so it is not easy to explain how to reproduce it.

**Comment by Sopor** [comment]
Here is one user that suddenly disappeared

2022-08-01 @ 23:01 the user got disconnected (i assume it was "An existing connection was forcibly closed by the remote host") because the user logged in again 23:01 or i should say that he tried to login again but the account was gone.
```
[2022-07-22 17:27] registered new User

[2022-08-01 23:01] [ USER LOGOUT   ]

[2022-08-01 23:01] [ FAILED AUTHENTICATION ]  |  Reason:  Registered users only.
```

**Comment by Kungen123** [comment]
yes this is a big problem when we lose accounts

**Comment by ryehamstrawberry** [comment]
> Unfortunately i don't want to be able to reproduce this. If anyone knows a way to reproduce it please post.

- Run a copy of Luadch v2.24 RC2/RC3 (note: I run pm2offliners and freshstuff scripts but neither of those should matter).
- Register an account then have someone login under the new credentials.
- Have the new account remain online for a time period while performing some activities such as browse filelists, searches, or download and allow other hub activity to occur (other join/parts, chat).
- Log out of the new account and stay offline for 24 hours (although in nearly all cases I have seen accounts disappear sooner).
- Ensure normal hub activity continues while the new account is offline (e.g. chat, downloads, users joining/parting).
- Do NOT register or make any other account modifications to any other accounts (no username/password changes, no delreg, no bans, etc.). The last admin activity should be the creation of the new account.
- Delete account and restart process if account does not disappear on its own at the end of the 24 hour period. I'm fairly confident that the bug will eventually reproduce itself under these conditions, but if anyone else is able to narrow down the possible variables causing this issue please speak up.

These are roughly the circumstances I've found the bug to most often occur, which makes me think that only the last account modified (whether its a new account created, deleted, banned, or changed) is affected. Although I have not confirmed this I think it's a good place to start.

**Comment by Toansy** [comment]
Yes this is a big problem. We are losing accounts

**Comment by corruptview** [comment]
It sure is

**Comment by ryehamstrawberry** [comment]
> Unfortunately i don't want to be able to reproduce this. If anyone knows a way to reproduce it please post.

There is much evidence this bug is affecting many users.

Have you tried following my advice in order to reproduce the bug?

**Comment by Sopor** [comment]
It seems that we have found something that can be useful to find what is going on.

I got a report that a user was deleted from a hub when a new month started and if you look at mine report it was also on the 1st: `[2022-08-01 23:01] [ FAILED AUTHENTICATION ]  |  Reason:  Registered users only.`.

So, be extra observant at the end of the month.

**Comment by Sopor** [comment]
### Now i'm 100 percent sure accounts suddenly disappear.

The account was created:
```
[2024-01-14 - 22:05:54] <Security]> [ REG ]--> User: [Owner]Xxxxx  |  registered new User: [100]Uuuuu  |  Level: 20 [ REG ]  |  Comment: no comment defined
```

The user has been logged in and logged out several times since that day, but today he couldn't login.

The last time the user was logged in and out successfully was earlier the same day the account disappeared:
```
[24-01-23] [06:20:20] <Security]> [ USER LOGIN       ]--> USER:  [100]Uuuuu |  IP:  123.123.123.123  |  CLIENT:  AirDC++\s4.21  |  SHARE: 3363097784849  |  CID:  XXXXXXXXXXXXXXXXXXXXXXXXXXX
```
```
[24-01-23] [06:23:03] <Security]> [ USER LOGOUT   ]--> USER:  [100]Uuuuu  |  IP:  123.123.123.123  |  CLIENT:  AirDC++\s4.21  |  SHARE: 3363097784849  |  CID:  XXXXXXXXXXXXXXXXXXXXXXXXXXX
```

In the OpChat i could read this:
```
[24-01-23] [18:37:30] <Security]> [ FAILED AUTHENTICATION ]--> User:  [100]Uuuuu  |  IP:  123.123.123.123  |  CID:  XXXXXXXXXXXXXXXXXXXXXXXXXX  |  Reason:  Registered users only.
```

If i then look in the `user.tbl` i can't find this users account, but if i look in `user.tbl.bak` i could.

At the bottom line i could read this:
```
    { by = "[Owner]Xxxxx", date = "2024-01-14 / 22:05:34", level = 20, nick = "[100]Uuuuu", password = "0th4beYb9o203E0Q0A5A", },
```

If you then look at the date and time of the `user.tbl.bak` you can see that it was the same date and time (plus and minus a few seconds because the time isn't synced) as the account was created:
![image](https://github.com/luadch/luadch/assets/5789283/b262dcc5-7695-48ae-ab4f-173a6262d16e)

### This bug really need to be fixed ASAP!

**Comment by Sopor** [comment]
The user also got an account in another hub the same day and if i look in the `user.tbl` and `user.tbl.bak` it looks like this after the user logged in.

**user.tbl.bak:**
```
    { by = "Xxxxx", date = "2024-01-14 / 23:31:42", level = 20, nick = "Uuuuu", password = "sclEE0thIvPHAFjd", },
```

**user.tbl:**
```
    { badpassword = 0, by = "Xxxxx", date = "2024-01-14 / 23:31:42", is_online = 1, lastconnect = 20240123195923, lastlogout = 20240123183911, lastseen = 20240123231146, level = 20, nick = "Uuuuu", password = "sclEE0thIvPHAFjd", },
```

![image](https://github.com/luadch/luadch/assets/5789283/963379bd-7c31-4d2b-9d5e-041cfec7ce14)

As you can see here above, this is how it should look like.

I don't know why users sometimes get deleted and sometimes not when creating a new account.

**Comment by ryehamstrawberry** [comment]
huge functionality crippling bug going on 2 years ignored ...

**Comment by Sopor** [comment]
Yes, it is very frustrating that this can't be fixed and that both @pulsar-de and @blasti doesn't reply. 😟

If it can't be recreated it must be possible to check the code (it started in 2.24) or create a debug version that can log everything that happens with the user.tbl and user.tbl.bak.

It seems that 2.23 works fine or at least it doesn't have this account issue.

A work around is to run `+reload` after every command `+reg`, `+delreg` or `+ban`.

**Comment by ryehamstrawberry** [comment]
Sopor, can you please confirm/deny whether the [steps I listed previously](https://github.com/luadch/luadch/issues/189#issuecomment-1219691805) allow you to reproduce the bug?

If we have confirmation that the bug is reproducible starting with a fresh v2.24 RC4 install, then hopefully @pulsar-de and @blasti can reproduce the bug too.

**Comment by maltenizer** [comment]
We've seen it happen too.

**Comment by blasti** [comment]
I looked briefly into the source.

1.  There are dozens and occasions of places where the user.tbl gets overwritten. Basically everytime a reg user logs in or out, the user.tbl is updated. 
2. There are several functions which are used for saving the user.tbl. I found at least two: `cfg.saveusers` and `util.savearray`.
3. I found caching stuff like `local user_tbl = hub.getregusers()` which might cause problems
4. I'm not sure how the backup stuff works right now in cfg.saveusers.
5. There are many places, where potential errors during writing the user.tbl file are not logged.
6. There are scripts, which delete users automatically (usercleaner, botcleaner).

Very hard to reason about. 

Did anybody found something in the error.log?

**Comment by Sopor** [comment]
@blasti Nothing is showing in the error.log.

The only thing we all know, is that it started in 2.24 after @pulsar-de made a lot of changes in the code.

As i have explained before, you can use `+reload` but that is only a work-around and not a fix.

**Comment by Sopor** [comment]
After some reports from other hub owners this issue seems to be in earlier versions before 2.24 too.

I have explained to @blasti what i have found out what could be wrong, but when @pulsar-de shines with its absence i don't know if this will never be fixed.

It is almost as this project is dead and that is very sad if that happens 😢

**Comment by Sopor** [comment]
When will you fix this @pulsar-de or @blasti ?

I'm pretty sure it is the `An existing connection was forcibly closed by the remote host.`.

If you get that before you have made a logout after you created a new account the account will disappear.
